File: csharp/Service.cs
using Aurelion.Connector.Core;

namespace Aurelion.Connector;

public sealed class Service : IOperationExecutor
{
    private readonly Backend _backend;
    private readonly Dictionary<string, Func<Dictionary<string, object?>, bool, string?, OperationResult>> _registry;

    public Service(string? dbPath, bool autoSeedMock = true)
    {
        _backend = new Backend(dbPath);
        _backend.InitDb(autoSeedMock);
        _registry = new()
        {
            ["create_account"] = CreateAccount,
            ["delete_account"] = DeleteAccount,
            ["list_accounts"] = ListAccountsOp,
            ["list_roles"] = ListRolesOp,
            ["list_privileges"] = ListPrivilegesOp,
            ["list_persons"] = ListPersonsOp,
            ["list_employments"] = ListEmploymentsOp,
        };
    }

    public OperationResult Execute(string operation, Dictionary<string, object?> payload,
                                    bool resultStorageRequested, string? correlationId = null)
    {
        if (!_registry.TryGetValue(operation, out var handler))
            throw new ArgumentException($"unsupported operation: '{operation}'");
        return handler(payload, resultStorageRequested, correlationId);
    }

    private OperationResult CreateAccount(Dictionary<string, object?> payload, bool _, string? __)
    {
        var username = AsString(payload, "username");
        var email = AsString(payload, "email");
        if (string.IsNullOrWhiteSpace(username)) throw new ArgumentException("username is required");
        if (string.IsNullOrWhiteSpace(email)) throw new ArgumentException("email is required");

        _backend.InsertAccount(username, email);
        return new(new Dictionary<string, object?> { ["username"] = username, ["email"] = email }, null);
    }

    private OperationResult DeleteAccount(Dictionary<string, object?> payload, bool _, string? __)
    {
        var username = AsString(payload, "username");
        if (string.IsNullOrWhiteSpace(username)) throw new ArgumentException("username is required");

        _backend.DeleteAccount(username);
        return new(new Dictionary<string, object?> { ["username"] = username }, null);
    }

    private OperationResult ListRecords(
        string datasetType,
        List<Dictionary<string, object?>> records,
        bool resultStorageRequested,
        string? correlationId)
    {
        if (resultStorageRequested)
            return new(null, Storage.WriteRecords(datasetType, records, correlationId));
        return new(records, null);
    }

    private OperationResult ListAccountsOp(Dictionary<string, object?> _, bool rsr, string? cid)
        => ListRecords("accounts", _backend.ListAccounts(), rsr, cid);

    private OperationResult ListRolesOp(Dictionary<string, object?> _, bool rsr, string? cid)
        => ListRecords("roles", _backend.ListRoles(), rsr, cid);

    private OperationResult ListPrivilegesOp(Dictionary<string, object?> _, bool rsr, string? cid)
        => ListRecords("privileges", _backend.ListPrivileges(), rsr, cid);

    private OperationResult ListPersonsOp(Dictionary<string, object?> _, bool rsr, string? cid)
        => ListRecords("persons", _backend.ListPersons(), rsr, cid);

    private OperationResult ListEmploymentsOp(Dictionary<string, object?> _, bool rsr, string? cid)
        => ListRecords("employments", _backend.ListEmployments(), rsr, cid);

    private static string? AsString(Dictionary<string, object?> map, string key)
        => map.TryGetValue(key, out var val) && val is string s ? s : null;
}

File: typescript/src/service.ts
import {
  deleteAccount,
  initDb,
  insertAccount,
  listAccounts,
  listEmployments,
  listPersons,
  listPrivileges,
  listRoles,
} from './backend.js';
import { writeRecords } from './core/storage.js';
import type { OperationExecutor, OperationResult } from './core/types.js';

type Handler = (
  payload: Record<string, unknown>,
  resultStorageRequested: boolean,
  correlationId: string | null,
) => OperationResult;

export class Service implements OperationExecutor {
  private readonly dbPath: string | undefined;
  private readonly registry: Map<string, Handler>;

  constructor(dbPath?: string, autoSeedMock = true) {
    this.dbPath = dbPath;
    initDb(dbPath, autoSeedMock);
    this.registry = new Map<string, Handler>([
      ['create_account', this.createAccount.bind(this)],
      ['delete_account', this.deleteAccountOp.bind(this)],
      ['list_accounts', this.listAccountsOp.bind(this)],
      ['list_roles', this.listRolesOp.bind(this)],
      ['list_privileges', this.listPrivilegesOp.bind(this)],
      ['list_persons', this.listPersonsOp.bind(this)],
      ['list_employments', this.listEmploymentsOp.bind(this)],
    ]);
  }

  execute(
    operation: string,
    payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null = null,
  ): OperationResult {
    const handler = this.registry.get(operation);
    if (!handler) throw new Error(`unsupported operation: '${operation}'`);
    return handler(payload, resultStorageRequested, correlationId);
  }

  private createAccount(payload: Record<string, unknown>): OperationResult {
    const username = payload['username'];
    const email = payload['email'];
    if (typeof username !== 'string' || !username) throw new Error('username is required');
    if (typeof email !== 'string' || !email) throw new Error('email is required');

    insertAccount(username, email, this.dbPath);
    return { payload: { username, email }, storageRef: null };
  }

  private deleteAccountOp(payload: Record<string, unknown>): OperationResult {
    const username = payload['username'];
    if (typeof username !== 'string' || !username) throw new Error('username is required');

    deleteAccount(username, this.dbPath);
    return { payload: { username }, storageRef: null };
  }

  private listRecords(
    datasetType: string,
    records: Record<string, unknown>[],
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    if (resultStorageRequested) {
      return {
        payload: null,
        storageRef: writeRecords(datasetType, records, correlationId),
      };
    }
    return { payload: records, storageRef: null };
  }

  private listAccountsOp(
    _payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    return this.listRecords('accounts', listAccounts(this.dbPath), resultStorageRequested, correlationId);
  }

  private listRolesOp(
    _payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    return this.listRecords('roles', listRoles(this.dbPath), resultStorageRequested, correlationId);
  }

  private listPrivilegesOp(
    _payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    return this.listRecords('privileges', listPrivileges(this.dbPath), resultStorageRequested, correlationId);
  }

  private listPersonsOp(
    _payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    return this.listRecords('persons', listPersons(this.dbPath), resultStorageRequested, correlationId);
  }

  private listEmploymentsOp(
    _payload: Record<string, unknown>,
    resultStorageRequested: boolean,
    correlationId: string | null,
  ): OperationResult {
    return this.listRecords('employments', listEmployments(this.dbPath), resultStorageRequested, correlationId);
  }
}
